README.txt
Introduction
============

This package contains an command line tool to the sealjet PDM
system.

Installation
------------

Install the package as usual.  Preferred installation method is
to add the package to the sealjet PDM main buildout.  Once installed,
teh setup will create a `sealjetpdm` executable/script.

Usage
-----

The `sealjetpdm` tool's usage is as follows::

    usage: sealjetpdm [-h] [-d] [-l LOGFILE] -u USERNAME -p PASSWORD -U URL
                      {import,query-import,lock-design,export,unlock-design} ...

    positional arguments:
      {import,query-import,lock-design,export,unlock-design}

    optional arguments:
      -h, --help            show this help message and exit
      -d, --debug
      -l LOGFILE, --logfile LOGFILE
      -u USERNAME, --username USERNAME
                            the user name to use
      -p PASSWORD, --password PASSWORD
                            the password
      -U URL, --url URL     the base URL to the sealjet PDM system's root cabinet

There are several subcommands:

- **query-import**
- **import**
- **export**
- **lock-design**
- **unlock-design**

Each of the subcommands has it's own help accessible using the `-h` option.


docs/HISTORY.txt
Changelog
=========

0.1 2011-04-28
--------------

- Initial release


sealjet/tool/api.py
# -*- coding: utf-8 -*-
#
# File: api.py
#
# Copyright (c) nexiles GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__    = """Stefan Eletzhofer <se@nexiles.de>"""
__docformat__ = 'plaintext'

import os
import json
import urllib
import logging
import httplib2

logger = logging.getLogger("sealjet.tool")

def q(v, *a, **kw):
    if type(v) == unicode:
        v = v.encode("utf-8")
    return urllib.quote_plus(v, *a, **kw)

def safe_unicode(v):
    if type(v) == str:
        return v
    elif type(v) == unicode:
        return v
    else:
        return str(v)

class API(object):

    def __init__(self, url, username, password):
        self.username = username
        self.password = password
        self.url = url

        self.http = httplib2.Http()

        self.plone_login()

        # default headers
        self.headers = {
                "Cookie": self.auth_cookie
        }

    def plone_login(self):
        url = self.url + "/login_form"
        headers =  {"User-agent" : "Mozilla/4.0 (compatible; MSIE 5.5; Windows NT)",
                    "Content-type": "application/x-www-form-urlencoded",
                    "Accept": "text/plain"}
        params = {'came_from': url,
                  'form.submitted': '1',
                  'js_enabled': '0',
                  'cookies_enabled': '',
                  'login_name': '',
                  'pwd_empty': '0',
                  '__ac_name': self.username,
                  '__ac_password': self.password,
                  'submit': 'Log in'}
        resp, content = self.http.request(url, "POST", body=urllib.urlencode(params), headers=headers)
        self.auth_cookie = resp['set-cookie']

    def request(self, api_method, body=None, **kwargs):
        logger.debug("REQUEST: %s %s" % (api_method, kwargs))
        headers = dict()
        headers.update(self.headers)

        if body:
            headers["Content-Length"] = str(len(body))
        else:
            headers["Content-Length"] = "0"

        params = ["m=%s" % api_method, ]
        for key in sorted(kwargs.keys()):
            value = kwargs[key]
            value = safe_unicode(value)
            params.append("%s=%s" % (key, q(value)))

        path = "@@API?%s" % "&".join(params)
        url = os.path.join(self.url, path)

        logger.debug("path=%s" % path)
        logger.debug("url=%s" % url)

        response, content = self.http.request(
                url,
                "GET",
                body,
                headers)

        logger.debug("RESPONSE: %s content: %d bytes." % (response.status, len(content)))
        logger.debug("RESPONSE: %r" % response)
        logger.debug("CONTENT : %r" % content)
        return response, content

    def query_import(self, username, partnumber):
        response, content = self.request("query_import", username=username, partnumber=partnumber)
        assert response.status == 200
        return json.loads(content)

    def capsule_import(self, username, partnumber):
        response, content = self.request("import", username=username, partnumber=partnumber)
        assert response.status == 200
        return json.loads(content)

    def capsule_export(self, username, partnumber, version=None):
        if version:
            response, content = self.request("export", username=username, partnumber=partnumber)
        else:
            response, content = self.request("export", username=username, partnumber=partnumber, version=version)
        assert response.status == 200
        return json.loads(content)

    def lock_design(self, username, partnumber):
        response, content = self.request("lock_design", username=username, partnumber=partnumber)
        assert response.status == 200
        return json.loads(content)

    def unlock_design(self, username, partnumber):
        response, content = self.request("unlock_design", username=username, partnumber=partnumber)
        assert response.status == 200
        return json.loads(content)

# vim: set ft=python ts=4 sw=4 expandtab :


sealjet/tool/__init__.py
# -*- coding: utf-8 -*-
#
# File: __init__.py
#
# Copyright (c) nexiles GmbH
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__author__    = """Stefan Eletzhofer <se@nexiles.de>"""
__docformat__ = 'plaintext'

import logging
import argparse
import urlparse

from sealjet.tool.api import API

logger = logging.getLogger("sealjet.tool")

def setup_logging(log_path="sealjet_tool.log", level=logging.DEBUG):
    logging.basicConfig(
            filename=log_path,
            level=level,
            format="%(asctime)s: %(name)s: %(levelname)s: %(message)s"
            )

def is_url(string):
    scheme, netloc, path, query, fragment = urlparse.urlsplit(string)
    if not scheme:
        raise argparse.ArgumentTypeError("'%s' does not look like an URL." % string)
    if query or fragment:
        raise argparse.ArgumentTypeError("'%s' contains query args or a fragment part -- please remove these." % string)
    return string


def do_query_import(api, args):
    logger.info("query_import: username=%s partnumber=%s" % (args.username, args.partnumber))
    print api.query_import(args.username, args.partnumber)

def do_import(api, args):
    print api.capsule_import(args.username, args.partnumber)

def do_export(api, args):
    print api.capsule_export(args.username, args.partnumber, args.version)

def do_unlock_design(api, args):
    print api.unlock_design(args.username, args.partnumber)

def do_lock_design(api, args):
    print api.lock_design(args.username, args.partnumber)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-d", "--debug", dest="loglevel", action="store_const", const=logging.DEBUG, default=logging.INFO)
    parser.add_argument("-l", "--logfile", default="sealjetpdm.log")
    parser.add_argument("-u", "--username", required=True, help="the user name to use")
    parser.add_argument("-p", "--password", required=True, help="the password")
    parser.add_argument("-U", "--url", required=True, type=is_url, help="the base URL to the sealjet PDM system's root cabinet")

    subparsers = parser.add_subparsers()
    
    parser_query_import = subparsers.add_parser("query-import")
    #parser_query_import.add_argument("username")
    parser_query_import.add_argument("partnumber")
    parser_query_import.set_defaults(command=do_query_import)

    parser_import = subparsers.add_parser("import")
    #parser_import.add_argument("username")
    parser_import.add_argument("partnumber")
    parser_import.set_defaults(command=do_import)

    parser_export = subparsers.add_parser("export")
    #parser_export.add_argument("username")
    parser_export.add_argument("partnumber")
    parser_export.add_argument("-V", "--version", dest="version")
    parser_export.set_defaults(command=do_export)

    parser_unlock_design = subparsers.add_parser("unlock-design")
    #parser_unlock_design.add_argument("username")
    parser_unlock_design.add_argument("partnumber")
    parser_unlock_design.set_defaults(command=do_unlock_design)

    parser_lock_design = subparsers.add_parser("lock-design")
    #parser_lock_design.add_argument("username")
    parser_lock_design.add_argument("partnumber")
    parser_lock_design.set_defaults(command=do_unlock_design)

    args = parser.parse_args()
    setup_logging(log_path=args.logfile, level=args.loglevel)

    logger.debug("-"*70)
    logger.debug("Start")

    api = API(args.url, args.username, args.password)
    args.command(api, args)

    logger.debug("Stop")
    logger.debug("-"*70)

if __name__ == "__main__":
    import doctest
    doctest.testmod()

# vim: set ft=python ts=4 sw=4 expandtab :
